Add String.{edit_distance,spellcheck}, UTF-8 aware spellchecking functions

dbuenzli opened the pull request:
*Note to reviewers, there's quite a bit of stdlib related PR's these days I don't mind if this doesn't get reviewed for 5.4, I did it now because I had to go over this again and the material is in my mind.*

This PR adds two new functions to the String module for devising “did you mean?” spell-checking functions on UTF-8 text:

```ocaml
val String.edit_distance : ?limit:int -> string -> string -> int 
val String.spellcheck : ?max_dist:(string -> int) -> string list -> string -> string list
```

Almost any program which deals with user provided input that needs to be matched against known strings can have its usability improved on errors by a simple mindless call to `String.spellcheck`, so it feels worth having them at hand in the Stdlib with good defaults.

The OCaml compiler itself has a form of these functions in [`Misc.edit_distance`] and [`Misc.spellcheck`]. This PR makes sure that at least the `Misc.edit_function` function can be replaced by the one proposed here. This in turn can fix the (unreported AFAIK) bug that these functions do not work meaningfully with the new [modest support for Unicode letters in
identifiers](https://github.com/ocaml/ocaml/pull/12664).


[`Misc.edit_distance`]: https://github.com/ocaml/ocaml/blob/trunk/utils/misc.ml#L941

[`Misc.spellcheck`]: https://github.com/ocaml/ocaml/blob/trunk/utils/misc.ml#L986

# `String.edit_distance` 

Like `Misc.edit_distance`, we use the [optimal string alignement distance][OSA] except we do it on arrays of `Uchar.t`. In contrast to the Levenshtein distance, OSA allows two letter transpositions to be counted as 1 edit rather than 2. In contrast to Damerau-Levenshtein, OSA does not allow the same subsequence to be modified twice. We do not compute Damerau-Levenshtein distance as this requires a counter per letter occuring in the string which is a bit more costly and unconvenient given that, we work on `Uchar.t` letters (vs. say bytes), the data structures we have in the Stdlib, and the module dependency constraints.

Like in `Misc.edit_distance` we add a `limit` parameter to limit the computation to a distance threshold. In this case only the values on a stripe of length `2*limit+1` around the diagonal of the dynamic programming matrix need to be computed. This significantly reduces the work done when you are interested in selecting strings that are not further away than a given limit which is usually the case in spellchecking tasks.

Unlike `Misc.edit_distance`, the proposed function has an unconditional allocation cost due to the internal conversion to `Uchar.t` arrays on which the distance is effectively computed. To compensate this we reduce the dynamic programming matrix as is often suggested but not done in `Misc.edit_distance` (you only need a ring buffer of 3 rows) .

Since the raw implementation result may be a bit difficult to understand, its implementation is introduced by a succession of three commits.

1. The first commit adds a classical OSA distance computation between two UTF-8 strings. It should be easy to match this implementation on pseudo code or the recurrence definition of the dynamic programming matrix that can be found e.g. [on wikipedia][OSA]. By looking at the indexation operations in the loop it is also easy to see that you don't need the full dynamic programming matrix but only a ring buffer of three rows.
   
2. The second commit makes the optimisation of only keeping a ring buffer of three rows of the smallest string. This makes the space requirements of the dynamic programming matrix `3*length(s)` with `s` the smallest string rather than `length(s0)*length(s1)`.
   
3. The third commit adds a `limit` parameter to stop the computation once a maximal distance is reached. This involves computing the dynamic programming array only on a stripe of length `2*limit+1` around the diagonal and compute the minimal in a row for early exit (not present in `Misc.edit_distance`). This makes the algorithm `O(limit*length(s))` time.

[OSA]: https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance#Optimal_string_alignment_distance

# `String.spellcheck`

This is a ready made spell-checking function that can be used with small dictionaries represented by a list of string for correcting user input.  Given a string `s` to correct it simply returns the list of words in the dictionary whose edit distance is the smallest and at most at a user provided `max_dist s` value. Note that this favours corrections to shorter words (this can be disputed but the aim here is to have an easy function one can quickly reach for. More elaborate schemes, e.g. based on word probability, can be devised by using the `String.edit_distance` primitive).

The `max_dist` function ~~defaults to `Fun.const 2`~~ (this is no longer the case see discussion below). This default can be discussed, especially on small strings. The problem is that it likely depends on application usage. For example `Misc.spellcheck` adapts the distance according to identifiers length. I would have liked to find some pragmatic user study for a good default, but except for the internet rumor that users make generally no more than 2 of edit errors per word, nothing really came out. My aim with the current definition is that programmers can simply use the function without thinking and have a good enougth and easy to predict default behaviour. I mention in the docs that the default is subject to change, in case we do want to change it later (e.g. make it more like `Misc.spellcheck`).

# Comments 

- Note that in contrast to `Misc.edit_distance` which is able to bail out immediately on large string differences, `String.edit_distance` has to create the `Uchar.t` arrays for the strings before being able to bail out on that check. This has an impact on the gc behaviour of `Misc.spellcheck`'s performance if `env` is large.

- It would have been nice to define `edit_distance` in the API of `Array` but we use it on subarrays here and we rather do not do an additional copy. This would have worked on `Dynarray` but I didn't want to introduce a dependency of `String` on `Dynarray` (though given that `Dynarray` are good `Buffer.t` for `Uchar.t` that's not entirely absurd). I'm willing to do this change if people think there's interest (i.e. have `Dynarray.edit_distance ?equal:('a -> 'a -> bool) -> ?limit:int -> 'a t -> 'a t -> int`) but I rather do the factorization in a subsequent PR.
        
- This PR adds a dependencies to `String` on the `Int`, `List` and `Array` modules (and `Uchar` but that was already a transitive dependency).  Any of these dependencies can be removed (some at the cost of code repetition) Please tell me if there's the desire to trim any of these.

dbuenzli committed:
Add String.edit_distance with a classic implementation.

dbuenzli committed:
array.ml: remove dependency on String module.

dbuenzli committed:
String.edit_distance: use a ring buffer of rows instead of a full matrix.

dbuenzli committed:
String.edit_distance: add [limit] to bound distance computation.

dbuenzli committed:
Add String.spellcheck.

wikku commented on stdlib/string.ml:
```diff
@@ -280,3 +280,70 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+  if Int.abs (len1 - len0) > limit then limit else
+  let s0, s1 = if len0 > len1 then s0, s1 else s1, s0 in
+  let len0, len1 = if len0 > len1 then len0, len1 else len1, len0 in
+  let rec loop row_minus2 row_minus1 row i len0 limit =
```
Why are `len0` and `limit` parameters? Also, why is `len1` recomputed below?

dbuenzli replied:
Because depending on how this gets compiled closure allocation may be avoided, ~~`len1` is not recomputed, it is accessed~~. See [this discussion](https://discuss.ocaml.org/t/tail-recursive-loops-and-closures/10845).

wikku replied:
`s0` and `s1` are accessed more often than `len0`, so why not pass them explicitly as well?

dbuenzli replied:
You are right. Simply because… I forgot! Though at that point we likely get to many vars for that manual paramter passing to be useful I should check goldbolt.

wikku commented:
I'm not sure if it works with row-by-row iteration and distance functions with transpositions, but edit distance can be computed in O(nd), where d is the result, by starting with a stripe of length 1 and widening it as the result at the diagonal increases.

But maybe that's not necessary, as O(nd) can be achieved in a black box manner by starting with limit 1 and doubling it until the result no longer increases.

gasche commented:
Thanks for the very complete, very careful PR description!

I haven't thought about it deeply yet, but here is a first round of feedback.

1. I wish that more tools had this "did you mean?" behavior on typos, so I support making it easy for other tool authors (using OCaml) to adopt. This being said, this could plausibly be done through a small opam package.

2. It is natural to ask whether the new stdlib support could be used in the compiler, replacing the code in Misc which could then be removed. (I suppose the motivation is also partly to have the compiler better support its new unicode identifiers; this interacts with (1) as it requires a compiler-libs or stdlib solution rather than a third-party opam package.) You probably have not done this, use your new stdlib code in the compiler, because it would require a bootstrap, which is more work than it should. (It would be fine to wait for the next OCaml release to do this change, for example.) But still, *would* we want to use your versions in the compiler? I would be nervous about performance, because we may run `spellcheck` on fairly large environments (set of defined names).

3. You mention that you haven't implemented an early bailout if the two strings have very-different sizes, which Misc has. Looking at the code, I see that you decode the string into a uchar array and compute the size at the same time. But you could also start by computing the size, to get the early bailout. My *intuition* (but I don't remember testing this in practice) is that it makes a performance difference for compiler-like use-case, where the environments can be large but most candidates bail out quickly. (Note: it's even possible to iterate on both strings in lockstep, to stop at size `s + cutoff` where `s` is the smallest of the two strings. I guess it is overkill, as iterating on strings to compute their uchar count is fast in practice.)

4. I wondered if it would be possible to take a `fold:((string -> int -> int) -> int -> int)` parameter instead of a `string list` parameter, to not force users to accumulate the environment into a list before calling the function. I thought that the Misc version did this already (given that the Env module in the compiler only provides folds), but I went to have a look again and in fact we use just a list. So maybe a list is fine.

gasche commented:
Also: I would not have `limit` and `max_dist` as optional parameters, rather as mandatory parameters, with a recommendation to use `~max_dist:(fun _ -> 2)` in the documentation to reduce decision fatigue. I think that this is a rather important choice as it noticeably impacts performance, so I would rather ensure that the user thinks about it.

dbuenzli commented:
Thanks for the comments @gasche.

> 1. I wish that more tools had this "did you mean?" behavior on typos, so I support making it easy for other tool authors (using OCaml) to adopt. This being said, this could plausibly be done through a small opam package.

I would say that for those tools that are programmed in OCaml your wish and your suggestion are in contradiction :–)

> Also: I would not have `limit` and `max_dist` as optional parameters, rather as mandatory parameters, with a recommendation to use `~max_dist:(fun _ -> 2)` in the documentation to reduce decision fatigue.  I think that this is a rather important choice as it noticeably impacts performance, so I would rather ensure that the user thinks about it.

I don't think `limit` should be a required parameter otherwise it's weird if you really want to compute the edit distance of two strings. Regarding `max_dist` it has a good performance default and I really rather not bother programmers with that (or that they end up making an unfortunate ergonomic decision). So again I prefer to have a good default, once they see result they are not happy with, they can start micro-managing by either tweaking `max_dist` or by devising their own `spellcheck` with the `String.edit_distance` primitive.

> 2\. It is natural to ask whether the new stdlib support could be used in the compiler, replacing the code in Misc which could then be removed.

In fact I thought that the code in Misc could simply have a fast past for the early bail out on ASCII identifiers:

```ocaml
let edit_distance s0 s1 cutoff = 
  let is_ascii s = String.for_all Char.Ascii.is_valid s in
  if not (is_ascii s0 && is_ascii s1) 
  then String.edit_distance ~limit:cutoff s0 s1 else
  let len0 = String.length s0 and len1 = String.length s1 in
  let limit = Int.min (Int.max len0 len1) cutoff in
  if Int.abs (len1 - len0) > limit then limit else 
  String.edit_distance ~limit:cutoff s0 s1
```

For now I opted not to do this in `String.edit_distance` in order not to make cultural assumptions :–) But this decision can certainly be revised, we could also, as you suggest in point 3. simply try to compute the `UTF-8` length without decoding which may not be that different of the performance of `is_ascii`.

> 3\. You mention that you haven't implemented an early bailout if the two strings have very-different sizes, which Misc has.

Just to make things clear, I do this but the check is done after the strings have been decoded to `Uchar.t` arrays and thus allocation of the arrays has occurred. See previous point.

> 4\. I wondered if it would be possible to take a `fold:((string -> int -> int) -> int -> int)` parameter instead of a `string list` parameter,

I rather have a simple interface. Sophisticated users can reuse the nice primitive `String.edit_distance`. The code of `String.spellcheck` is rather straightforward (but I'd like to have it because the effort to think about it and to write it down may make programmers abandon the idea of implementing a “did you mean?”, I want “did you mean?” in the simplest OCaml scripts).

dbuenzli commented:
So just to give an idea I made a [toy benchmark] which computes the distance `hello` with limits 2,3,4 against the 235976 words of my `/usr/share/dict/words` (so these timing represent ~700'000 checks against purely ASCII words). 

Here are the numbers, the first run corresponds to the `String.edit_distance` of this PR (as found in `B0_std`, it's exactly the same code), the second uses a cut and paste of `Misc.edit_distance` and the third one is the function with the ASCII fast path mentioned in my previous message:

```
> hyperfine "$(brzo -b --path)"
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out
  Time (mean ± σ):     100.3 ms ±   0.5 ms    [User: 96.2 ms, System: 3.4 ms]
  Range (min … max):    99.6 ms … 101.9 ms    28 runs
 
> hyperfine "$(brzo -b --path) --use-misc"
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc
  Time (mean ± σ):      95.0 ms ±   0.6 ms    [User: 90.5 ms, System: 3.7 ms]
  Range (min … max):    93.9 ms …  96.5 ms    30 runs
 
> hyperfine "$(brzo -b --path) --use-ascii-path"
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii-path
  Time (mean ± σ):      87.6 ms ±   0.5 ms    [User: 83.5 ms, System: 3.4 ms]
  Range (min … max):    86.3 ms …  88.8 ms    33 runs
``` 

[toy benchmark]: https://gist.github.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161

dbuenzli commented:
And just to round that up. I tried to compute the `Uchar.t` length first for the early bail out (via [this patch](https://gist.github.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161#file-utf-8-len-patch)) but at least on the ASCII benchmark above that came out slower:

```
> hyperfine "$(brzo -b --path)"
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out
  Time (mean ± σ):     106.9 ms ±   0.6 ms    [User: 102.5 ms, System: 3.6 ms]
  Range (min … max):   105.3 ms … 107.8 ms    27 runs
```

gasche commented:
Thanks! Your benchmark computes edit distance three times. What is the performance of the 0-times version, that just does the rest of the logic? (This helps estimate relative differences in efficiency between the variants you measure.)

The ascii version is quite nice; today all identifiers are ascii so this is just fine, but I'm thinking than in the future it is more likely that the user-provided identifier may have unicode, while the environment will be mostly ascii-only. In the use-case of computing the edit distance of one string against many, one "needle" in a haystack of other strings, (one could use a clever automata, anyways) we could in fact compute the length of that needle once, which would provide a significant speedup in the utf8-length-bailout version. But this makes the interface slightly uglier (providing an optional parameter for the uchar length of each argument?), so I'm not too convinced myself that this is a good idea.

dbuenzli commented:
Here's the cost of just loading and iterating over the words of the dictionary:

```
> hyperfine "$(brzo -b --path) --just-iter"
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter
  Time (mean ± σ):      20.6 ms ±   0.3 ms    [User: 17.8 ms, System: 2.2 ms]
  Range (min … max):    20.1 ms …  21.7 ms    126 runs
```

gasche commented:
I'm satisfied with your benchmarking work: I think the ascii bailout is a reasonable approach, we could have it in Misc as your suggest or include it in Stdlib if you agree (this would have my mild preference, but no pressure).

Looking at the Misc implementation again, I see that the `max_dist` function we use (rephrased in terms of your API) is the following:

```ocaml
  let max_dist name =
    match String.length name with
      | 1 | 2 -> 0
      | 3 | 4 -> 1
      | 5 | 6 -> 2
      | _ -> 3
```

You propose `(fun _ -> 2)` instead by default. This seems to have worse behavior in the case of one-uchar strings, where all other one-uchar strings would be returned as suggestions (whereas Misc disables spell-checking entirely for those characters). If the intention is to have a good default for catching human-made typos, I would maybe suggest `(1 | 2) -> 0 | (3 | 4) -> 1 | _ -> 2`.

dbuenzli commented:
> we could have it in Misc as your suggest or include it in Stdlib if you agree (this would have my mild preference, but no pressure).

Why not. But I'd like to also benchmark with an international list of words.

> I would maybe suggest `(1 | 2) -> 0 | (3 | 4) -> 1 | _ -> 2`.

Yes. (I wanted to avoid to compute the Unicode length in the default but it's misguided, it's better to have this).

dbuenzli committed:
String.spellcheck: better default max_dist.

dbuenzli committed:
Address review comment from wikku.

dbuenzli commented:
So I changed the default of `String.spellcheck` as suggesed by @gasche. I will move and expose the `uchar_utf_8_decode_length_of_byte` function to `Uchar` in a subsequent PR, if this gets merged. It's a [useful](https://github.com/dbuenzli/jsont/blob/bd7c82888ba146c7d68a1216a69758277f96beed/src/bytesrw/jsont_bytesrw.ml#L9-L18) function to have in the stdlib.

Now regarding adding the optimization for ASCII by default I think I rather not. It's (rather obviously) slower as soon as there are characters beyond ASCII and the function advertises as working on Unicode, see numbers below.

Besides `String.edit_distance` is likely faster anyways than the current[^1] `Misc`. An empirical observation seems to indicate `Misc` becomes slower when the checked word reaches 8 characters. I suspect you do pay a lot for the full matrix when the words grow larger. For example testing the word `List.fald_left` against the `/usr/share/dict/words` (I tried to convince `ocp-index` to dump me all the qualified names it knows but I failed, if someone knows a trick, let me know) we get:

[^1]: It would be of course be easy to make it faster than what is proposed here, I assume we are interested in not having a regression with the status quo.

```
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-proposal --word List.fald_left
  Time (mean ± σ):     136.0 ms ±   0.7 ms    [User: 131.2 ms, System: 4.0 ms]
  Range (min … max):   134.3 ms … 138.3 ms    21 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii --word List.fald_left
  Time (mean ± σ):     123.4 ms ±   0.6 ms    [User: 118.6 ms, System: 4.0 ms]
  Range (min … max):   122.6 ms … 125.3 ms    23 runs
 
Benchmark 3: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc --word List.fald_left
  Time (mean ± σ):     228.1 ms ±   0.5 ms    [User: 221.5 ms, System: 5.5 ms]
  Range (min … max):   227.3 ms … 229.0 ms    12 runs
 
Benchmark 4: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter --word List.fald_left
  Time (mean ± σ):      20.5 ms ±   0.2 ms    [User: 17.8 ms, System: 2.3 ms]
  Range (min … max):    20.1 ms …  21.4 ms    128 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter --word List.fald_left ran
    6.01 ± 0.06 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii --word List.fald_left
    6.62 ± 0.07 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-proposal --word List.fald_left
   11.11 ± 0.10 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc --word List.fald_left
```

Since I found [a way](https://gist.github.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161#file-dump-aspell-dict-sh) to dump `aspell` dictionaries as word lists. I did a bit more testing on Unicode data. 

The following benchmark is done on the serbian dictionary which has the advantage of having both the cyrillic and latin script words of the language. This means we get a mix of words made of only ASCII characters, ASCII characters and multibyte characters and only multibyte character (686'453 words in total, the timings below are for ~2 million edit distance computations).  

Here are the results against a pure ASCII word (`funkcija`) followed by the same word in the cyrillic script (`функција`) (note that we are faster than `Misc`, but `Misc` is meaningless anyways on these):

```
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-proposal --dict-file /tmp/dict-sr --word funkcija
  Time (mean ± σ):     441.3 ms ±   2.0 ms    [User: 425.6 ms, System: 13.0 ms]
  Range (min … max):   437.1 ms … 444.9 ms    10 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii --dict-file /tmp/dict-sr --word funkcija
  Time (mean ± σ):     479.1 ms ±   0.9 ms    [User: 463.8 ms, System: 12.7 ms]
  Range (min … max):   478.1 ms … 481.0 ms    10 runs
 
Benchmark 3: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc --dict-file /tmp/dict-sr --word funkcija
  Time (mean ± σ):     476.3 ms ±   1.3 ms    [User: 460.9 ms, System: 12.3 ms]
  Range (min … max):   474.8 ms … 479.4 ms    10 runs
 
Benchmark 4: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter --dict-file /tmp/dict-sr --word funkcija
  Time (mean ± σ):      58.8 ms ±   0.3 ms    [User: 52.1 ms, System: 5.3 ms]
  Range (min … max):    58.4 ms …  60.6 ms    47 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter --dict-file /tmp/dict-sr --word funkcija ran
    7.50 ± 0.05 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-proposal --dict-file /tmp/dict-sr --word funkcija
    8.10 ± 0.05 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc --dict-file /tmp/dict-sr --word funkcija
    8.14 ± 0.05 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii --dict-file /tmp/dict-sr --word funkcija
```

```
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-proposal --dict-file /tmp/dict-sr --word функцијаb
  Time (mean ± σ):     470.7 ms ±   1.7 ms    [User: 455.3 ms, System: 13.2 ms]
  Range (min … max):   468.5 ms … 473.6 ms    10 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii --dict-file /tmp/dict-sr --word функција
  Time (mean ± σ):     480.7 ms ±   1.6 ms    [User: 464.3 ms, System: 13.7 ms]
  Range (min … max):   477.0 ms … 482.3 ms    10 runs
 
Benchmark 3: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc --dict-file /tmp/dict-sr --word функција
  Time (mean ± σ):     725.5 ms ±   1.3 ms    [User: 704.3 ms, System: 18.6 ms]
  Range (min … max):   723.6 ms … 727.6 ms    10 runs
 
Benchmark 4: /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter --dict-file /tmp/dict-sr --word функција
  Time (mean ± σ):      59.3 ms ±   0.5 ms    [User: 52.2 ms, System: 5.7 ms]
  Range (min … max):    58.4 ms …  60.8 ms    47 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --just-iter --dict-file /tmp/dict-sr --word функција ran
    7.94 ± 0.08 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-proposal --dict-file /tmp/dict-sr --word функција
    8.11 ± 0.08 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-ascii --dict-file /tmp/dict-sr --word функција
   12.24 ± 0.12 times faster than /Users/dbuenzli/tmp/eddist/_b0/brzo/ocaml-exec-native/a.out --use-misc --dict-file /tmp/dict-sr --word функција
```

dbuenzli committed:
String.edit_distance: add an obvious shortcut for limit <= 1.

This is mainly useful for String.spellcheck once it has reached
a minimal distance of 0.

dbuenzli commented:
Just FTR I have overcome my indexing fears and [have implemented](https://gist.github.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161#file-test_edit_distance-ml-L112-L176) in the benchmark a version which uses 3 buffers of `2 * limit + 2` length for the matrix.

FWIW it came out slower on the serbian dictionary benchmark; I suspect you pay for more branches and natural language words are rather small so the buffer sizes are not that different from those of the smaller word (however a function like `spellcheck` could reuse the buffers).

It's also strictly more index fiddly and less readable. It could make a difference on long identifiers though but for now I decided to leave things as they are. It seems the performance tradeoffs are quite data dependent.

gasche commented:
I'm convinced by your performance measurements and feel okay with letting you decide whether to have the fast path in the stdlib version or not -- you are the expert. I didn't mean to throw you into an infinite iterative-deepening benchmarking loop.

dbuenzli commented:
Oh don't worry that's self-imposed procrastination :–) I'd still like to run the bench on the dictionary of all qualified names of my opam switch, but that will have to wait a bit.

gasche commented:
(I thought that using this in utils/misc.ml would require a bootstrap, but in fact I was mistaken, I don't think we need a bootstrap to use the code in stdlib/foo.ml.)

dbuenzli commented:
So I have [a small script](https://gist.githubusercontent.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161/raw/2c12e1fa72c2aeacce1254f5d46186f4871b7aa5/ocp_index_dump.ml) to dump all the qualified identifiers of an opam switch with `ocp-index`. On a switch with ~100 packages which includes alternate standard libraries (except `base`, somehow `ocp-index` explodes when base is installed) that's about `427'504` ASCII identifiers (and thus the benchmark makes ~1.3Mio edit distance computation). 

Here again the results are quite data dependent, with short identifiers `Misc` is much faster. I computed the [histogram](https://gist.githubusercontent.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161/raw/2c12e1fa72c2aeacce1254f5d46186f4871b7aa5/dict-opam.hist) and [the quartiles](https://gist.githubusercontent.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161/raw/2c12e1fa72c2aeacce1254f5d46186f4871b7aa5/dict-opam.stats) of the length of identifiers and the first quartile (25%) comes out at length of 24. But I couldn't quickly figure out if in the compiler the edit distance is in fact really computed on fully qualified names.

There are a few things that could be tried (e.g. rather than benchmark `edit_distance` benchmark `spellcheck`, have the windowed version and have `spellcheck` reuse window buffers and the Uchar.t conversion of the word to correct) but if I don't get the time to get back to this also feel that this could be done later and the PR merged as is. Also given the absolute timing numbers and the number of edit distance computations this represents I would personally not be nervous about replacing `Misc.edit_distance` by `String.edit_distance`.

So here are a few results on the "opam dictionary", with `fald_left`, `List.fald_left` and an identifier that has the quartile length 24 `123456789012345678901234` (I added all the data and code mentioned in this PR in [the gist](https://gist.githubusercontent.com/dbuenzli/84be5567d4d3d6f987c18f72b9d2e161/) if people want to have a closer look).

```sh
> b0 -- bench-dict-opam --word=fald_left          
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-proposal --dict-file dict-opam --word=fald_left
  Time (mean ± σ):     335.9 ms ±   1.3 ms    [User: 320.2 ms, System: 13.9 ms]
  Range (min … max):   334.0 ms … 337.7 ms    10 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-window --dict-file dict-opam --word=fald_left
  Time (mean ± σ):     336.1 ms ±   1.5 ms    [User: 320.2 ms, System: 13.3 ms]
  Range (min … max):   334.3 ms … 338.8 ms    10 runs
 
Benchmark 3: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-ascii --dict-file dict-opam --word=fald_left
  Time (mean ± σ):     139.4 ms ±   0.4 ms    [User: 129.6 ms, System: 8.3 ms]
  Range (min … max):   138.9 ms … 140.6 ms    21 runs
 
Benchmark 4: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-misc --dict-file dict-opam --word=fald_left
  Time (mean ± σ):      70.5 ms ±   0.7 ms    [User: 62.5 ms, System: 6.6 ms]
  Range (min … max):    69.3 ms …  72.9 ms    41 runs
 
Benchmark 5: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --just-iter --dict-file dict-opam --word=fald_left
  Time (mean ± σ):      48.6 ms ±   0.8 ms    [User: 40.9 ms, System: 6.2 ms]
  Range (min … max):    47.4 ms …  50.9 ms    59 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --just-iter --dict-file dict-opam --word=fald_left ran
    1.45 ± 0.03 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-misc --dict-file dict-opam --word=fald_left
    2.87 ± 0.04 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-ascii --dict-file dict-opam --word=fald_left
    6.91 ± 0.11 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-proposal --dict-file dict-opam --word=fald_left
    6.91 ± 0.11 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-window --dict-file dict-opam --word=fald_left
```
```sh
> b0 -- bench-dict-opam --word=List.fald_left
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-proposal --dict-file dict-opam --word=List.fald_left
  Time (mean ± σ):     372.3 ms ±   1.0 ms    [User: 358.2 ms, System: 11.9 ms]
  Range (min … max):   370.5 ms … 373.3 ms    10 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-window --dict-file dict-opam --word=List.fald_left
  Time (mean ± σ):     368.6 ms ±   1.6 ms    [User: 355.1 ms, System: 11.3 ms]
  Range (min … max):   366.6 ms … 371.6 ms    10 runs
 
Benchmark 3: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-ascii --dict-file dict-opam --word=List.fald_left
  Time (mean ± σ):     172.8 ms ±   0.8 ms    [User: 162.9 ms, System: 8.3 ms]
  Range (min … max):   172.1 ms … 175.1 ms    16 runs
 
Benchmark 4: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-misc --dict-file dict-opam --word=List.fald_left
  Time (mean ± σ):     149.3 ms ±   0.5 ms    [User: 140.3 ms, System: 7.7 ms]
  Range (min … max):   148.4 ms … 150.3 ms    19 runs
 
Benchmark 5: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --just-iter --dict-file dict-opam --word=List.fald_left
  Time (mean ± σ):      48.3 ms ±   0.4 ms    [User: 40.9 ms, System: 6.1 ms]
  Range (min … max):    47.8 ms …  50.1 ms    59 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --just-iter --dict-file dict-opam --word=List.fald_left ran
    3.09 ± 0.03 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-misc --dict-file dict-opam --word=List.fald_left
    3.58 ± 0.03 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-ascii --dict-file dict-opam --word=List.fald_left
    7.63 ± 0.07 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-window --dict-file dict-opam --word=List.fald_left
    7.71 ± 0.07 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-proposal --dict-file dict-opam --word=List.fald_left
```

```sh
> b0 -- bench-dict-opam --word=123456789012345678901234
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-proposal --dict-file dict-opam --word=123456789012345678901234
  Time (mean ± σ):     432.3 ms ±  11.3 ms    [User: 417.1 ms, System: 13.0 ms]
  Range (min … max):   408.8 ms … 446.6 ms    10 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-window --dict-file dict-opam --word=123456789012345678901234
  Time (mean ± σ):     418.4 ms ±   8.4 ms    [User: 403.5 ms, System: 12.5 ms]
  Range (min … max):   402.4 ms … 429.4 ms    10 runs
 
Benchmark 3: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-ascii --dict-file dict-opam --word=123456789012345678901234
  Time (mean ± σ):     248.6 ms ±   2.6 ms    [User: 237.8 ms, System: 9.2 ms]
  Range (min … max):   244.6 ms … 251.1 ms    12 runs
 
Benchmark 4: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-misc --dict-file dict-opam --word=123456789012345678901234
  Time (mean ± σ):     367.3 ms ±   0.8 ms    [User: 353.0 ms, System: 12.4 ms]
  Range (min … max):   366.3 ms … 368.5 ms    10 runs
 
Benchmark 5: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --just-iter --dict-file dict-opam --word=123456789012345678901234
  Time (mean ± σ):      48.2 ms ±   0.5 ms    [User: 40.9 ms, System: 6.0 ms]
  Range (min … max):    47.4 ms …  50.2 ms    59 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --just-iter --dict-file dict-opam --word=123456789012345678901234 ran
    5.16 ± 0.07 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-ascii --dict-file dict-opam --word=123456789012345678901234
    7.62 ± 0.08 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-misc --dict-file dict-opam --word=123456789012345678901234
    8.68 ± 0.19 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-window --dict-file dict-opam --word=123456789012345678901234
    8.97 ± 0.25 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-edit-distance/test-edit-distance --use-proposal --dict-file dict-opam --word=123456789012345678901234
```

gasche commented:
I agree than taking less than half a second on all names exported by all OCaml compilation units in opam is no cause for concern. In your List.fald_left test, your code is about 3x slower than Misc (if we substract the base iteration time), only 1.3x slower with the ascii fast path, and this is an acceptable code to pay in exchange for the warm fuzzy feeling of handling all of unicode correctly.

gasche commented on stdlib/stringLabels.mli:
```diff
@@ -395,6 +395,48 @@ val is_valid_utf_16le : t -> bool
+    for the notion of character. Decoding errors are replaced by
+    {!Uchar.rep}. Normalizing the strings to
+    {{:https://unicode.org/glossary/#normalization_form_c}NFC} gives
+    better results.
```
I would be tempted to move this paragraph towards the end of the documentation string (maybe after the note), because I suspect that for many it will be a point of detail relative to the meaning of the `~limit` parameter below.

gasche reviewed (approved):
Approved. (This is a stdlib change so it will need a second maintainer approval.)

I have not in fact reviewed the core implementation in details, but I'm broadly confident and there is a nice testsuite.

Minor comment: currently you reuse the Misc interface relatively to the `limit` parameter, that you cap the return value with `limit`. I briefly wondered if it wouldn't be nicer to return `None` in that case (when we know that the distance is at least `limit`, but do not know for sure that it is `limit`). But this makes the API less pleasant to use for people who don't provide a limit, so I guess it's not such a good idea. Oh well...

dbuenzli committed:
Address doc review from Gabriel.

dbuenzli committed:
String.spellcheck: compute the word's uchars only once.

dbuenzli commented:
>  But this makes the API less pleasant to use for people who don't provide a limit, so I guess it's not such a good idea. Oh well...

Having `None` is also strictly more annoying if you are trying to compute minimums, see `String.spellcheck`.

In the end I couldn't get out of my mind that `spellcheck` was converting the spellchecked word to `uchars` for each word of the dictionary so in 70ae9eb742f421f42 I factored out this computation (a bench below, see the `reuse` timings, this is a use of `spellcheck` so it's more indicative of the time taken on a concrete dict than earlier benchmarks).

```
> b0 -- bench-spellcheck-dict-opam --word функција
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-proposal --dict-file dict-opam --word функција
  Time (mean ± σ):     148.1 ms ±   0.9 ms    [User: 139.5 ms, System: 7.0 ms]
  Range (min … max):   147.1 ms … 150.0 ms    19 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-reuse --dict-file dict-opam --word функција
  Time (mean ± σ):     110.9 ms ±   1.5 ms    [User: 103.1 ms, System: 6.4 ms]
  Range (min … max):   107.2 ms … 114.6 ms    26 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-reuse --dict-file dict-opam --word функција ran
    1.34 ± 0.02 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-proposal --dict-file dict-opam --word функција
> b0 -- bench-spellcheck-dict-sr --word функција
Benchmark 1: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-proposal --dict-file dict-sr --word функција
  Time (mean ± σ):     128.3 ms ±   0.6 ms    [User: 120.4 ms, System: 6.4 ms]
  Range (min … max):   127.4 ms … 129.8 ms    22 runs
 
Benchmark 2: /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-reuse --dict-file dict-sr --word функција
  Time (mean ± σ):     111.5 ms ±   0.9 ms    [User: 103.7 ms, System: 6.3 ms]
  Range (min … max):   110.8 ms … 115.4 ms    26 runs
 
Summary
  /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-reuse --dict-file dict-sr --word функција ran
    1.15 ± 0.01 times faster than /Users/dbuenzli/tmp/eddist/_b0/b/user/test-spellcheck/test-spellcheck --use-proposal --dict-file dict-sr --word функција
```

dbuenzli committed:
Add forgotten run of tools/sync_stdlib_docs.

nojb commented on stdlib/array.ml:
```diff
@@ -441,12 +441,10 @@ let fast_sort = stable_sort
+  invalid_arg
+    ("Array.shuffle: 'rand " ^ int (i + 1) ^
+     "' returned " ^ int j ^
+     ", out of expected range [0; " ^ int i ^ "]")
```
This change seems unnecessary. But OK to leave it in if you feel strongly about it :)

dbuenzli replied:
Not sure I understand, did I change something about the error message ? Otherwise I need this change: `Array` was depending on `String` only because of the `String.concat` there, with this PR `String` depends on `Array` so I need to break the loop.

nojb replied:
I had missed the technical point about breaking the dependency loop. Sorry for the noise.

gasche replied:
(A short comment `(* not using String.concat to avoid an Array->String dependency *)` would possibly have been helpful here.)

nojb commented on stdlib/string.ml:
```diff
@@ -280,3 +280,98 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+  done;
+  uchars, !k
+
+let _edit_distance ?(limit = Int.max_int) s (s0, len0) s1 =
```
Not asking for a change, but I would have avoided the `_` here which is normally used to identify unused functions.

nojb commented on stdlib/string.mli:
```diff
@@ -395,6 +395,48 @@ val is_valid_utf_16le : t -> bool
+    {{:https://unicode.org/glossary/#normalization_form_c}NFC} gives
+    better results.
+
+    {b Note.} This implements the simpler optimal string alignement
```
Nit, but I think adding a few capitals improves understanding of the comment (if you do not know what it is talking about beforehand).
```suggestion
    {b Note.} This implements the simpler Optimal String Alignement (OSA)
```

nojb commented on stdlib/string.ml:
```diff
@@ -280,3 +280,98 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+  | '\xC2' .. '\xDF' -> 2
+  | '\xE0' .. '\xEF' -> 3
+  | '\xF0' .. '\xF4' -> 4
+  | _ -> 0
```
Can we return `1` instead of `0` here and get rid of the `if` in `utf_8_char_length` below?

dbuenzli replied:
Rather not, after this PR I want to expose this in `Uchar`. Returning `0` means that you have an UTF-8 decoding error what you do on a decoding error is application dependent.

dbuenzli replied:
(That is if we return `1` we can no longer distinguish invalid UTF-8 starter bytes)

nojb replied:
OK!

nojb reviewed (approved):
LGTM (did not read the algorithm itself in detail).

Happy to see non-trivial functions enter the standard library!

dbuenzli committed:
Address review from Nicolás.

dbuenzli committed:
Update changes.

gasche commented:
We have two maintainers approval, but I wonder if @Octachron wants to have a look before we merge. (As our resident second unicode-and-also-edit-distances nerd, or in case he has opinions on dependencies among stdlib modules.) @Octachron, could you in the next few days tell us whether you want to have a look (then we can of course wait longer) or are happy to let us merge right now?

Octachron commented:
Yes, I am planning to have a look at the beginning of next week.

Octachron commented on stdlib/string.mli:
```diff
@@ -395,6 +395,48 @@ val is_valid_utf_16le : t -> bool
+*)
+
+val spellcheck :
+  ?max_dist:(string -> int) -> string list -> string -> string list
```
I am wondering if it would not be better to represent the dictionary as an iterator `(string -> unit) -> unit` rather than a list here.

Indeed, the case when the word belongs to the dictionary feels a bit strange with the current design: we end up traversing the whole list to build a singleton list containing our initial word.

This suggests to me that a common pattern would be to first test if the word belongs exactly to the dictionary (with a string set or a trie) before calling the `spellcheck` function. But then if the dictionary is not a list, it seems better to avoid constructing an intermediary list.

As a small positive side-effect, this would remove the dependency on the `List` module which feels quite accidental.

dbuenzli replied:
> Indeed, the case when the word belongs to the dictionary feels a bit strange with the current design: we end up traversing the whole list to build a singleton list containing our initial word.

Well usually when you get to use `spellcheck` you already know you have mismatch. That being said @gasche already suggested to have something more flexible than a list so I think I'm ok with making this slightly more complex for users in the name of flexibility. 

> As a small positive side-effect, this would remove the dependency on the `List` module which feels quite accidental.

I will still need `List.rev` :–)

Octachron replied:
I was thinking that the ordering is mostly cosmetics, but it is probably better to preserve a possible alphabetical order from the dictionary.

dbuenzli replied:
Yes especially because suggestions ("the dictionary") may not be in alphabetical order at all but sorted according to some notion of application defined preference.

dbuenzli replied:
`String.spellcheck`'s dictionaries are now represented by iterators.

Octachron commented on stdlib/string.ml:
```diff
@@ -280,3 +280,98 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+  let[@inline] minimum a b c = Int.min a (Int.min b c) in
+  let s1, len1 = uchar_array_of_utf_8_string s1 in
+  let limit = Int.min (Int.max len0 len1) limit in
+  if Int.abs (len1 - len0) > limit then limit else
```
`>` ⇒ `>=` since we need at least `abs (len1 - len0)` changes at cost one to align the sizes, and thus we can already short-circuit the computation when `abs (len1 - len0)` = `limit`.

dbuenzli replied:
Yes indeed!

dbuenzli committed:
Short path: improve bound check.

dbuenzli committed:
String.spellcheck: represent dictionaries by their iteration.

dbuenzli commented:
In [this comment](https://github.com/ocaml/ocaml/pull/13570#issuecomment-2648844087) @gasche proposed to unify the `uchar` length computations made there with the one done here. 

I just want to mention that the length computation function in this PR does not return correct results on invalid UTF-8 strings if they are decoded like `uchar_array_of_utf_8_string` does. The reason is that it skips over bytes, assuming a correct encoding (here's an example: it will return `1` on `"\xED\x0A\x20"` but a decode will result in two uchars:  `Uchar.rep` and `U+0020`).

I think this is good enough for spellcheck's default `max_dist` function because in practice you should not be giving invalid UTF-8 to spellcheck to these functions. If we want to switch to a function that returns the correct length on invalid UTF-8 strings that means we'll do the UTF-8 decoding twice (once in `max_dist` and then once via the call to `uchar_array_of_utf_8_string`).

Octachron commented on stdlib/string.ml:
```diff
@@ -280,3 +280,100 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+      let min = minimum
+          (row_minus1.(j-1) + cost) (* substitute *)
+          (row_minus1.(j) + 1)      (* delete *)
+          (row.(j-1) + 1)           (* insert *)
```
There is a bit of unnecessary work here.

 If we name `m` the edit distance matrix, we have by induction that
` m.(i).(j) >= abs (i -j)`. Thus outside of the stripe `i - limit + 1` and `i + limit - 1` , we know that distance is necessarily greater or equal to `limit`. 
However, at the lower boundary `j=i - limit + 1`, the read
```ocaml
row.(i - limit) + 1
```
is illegal if `i - limit > 0` because we have yet to write on this buffer location that was previously holding the value `m.(i-2).(i - limit)` and thus might be `limit - 2`. 

I can see two fixes, the explicit one
```
    if (i - limit) > 0 then row.(i - limit) <- limit;
    for j = Int.max 1 (i - limit + 1) to Int.min len1 (i + limit - 1) do
```
and the implicit/magical one:
```
    for j = Int.max 1 (i - limit) to Int.min len1 (i + limit - 1) do
```
which uses the fact that `m.(i-2).(i - limit - 1) + 1` is by induction greater than `limit`

dbuenzli replied:
What you write make sense but I find it difficult to write a bogus test case. Do you happen to have any ? or does the inductive propogation entail that the bogus read is harmless in the end ?

Octachron replied:
With the current slightly  larger than necessary stripe what happens is that the read is innocuous because at the boundary `j = i - limit - 1`  we read `m.(i).(j-1)` = `m.(i-2).(i - limit - 2)` which exists inside the buffer and is at least `limit` and thus everything works out.

The bug only rears its head if you try to reduce the stripe size to `i - limit +1, i - limit - 1` relying on the hypothesis above without noticing the problem with the read on the boundary. Then the testsuite triggers on the `d("fun", "function")` case.

dbuenzli replied:
Changed as you suggested and added a comment about the rogue read and why it works.

Octachron commented on stdlib/string.ml:
```diff
@@ -280,3 +280,100 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+    let len1 = Array.length row - 1 in
+    let row_min = ref Int.max_int in
+    row.(0) <- i;
+    for j = Int.max 1 (i - limit - 1) to Int.min len1 (i + limit + 1) do
```
There is an overflow issue with `js_of_ocaml` where  `Sys.max_string_length` is greater than `Int.max_size / 2`. However, it sounds sensible to raise an error in this case.

dbuenzli replied:
Are you suggesting something should be done ? That would show up if you start computing the (unlimited) edit distance between ~1GB strings without a `limit`. Not sure if that use case makes much sense.

Octachron replied:
Good question, I was thinking of raising `Invalid_argument "edit_distance does not support string larger than Int.max_int /2"` to avoid computing a wrong value in this corner case, but this puts quite a strain on usability to warn against a very narrow corner case.

dbuenzli replied:
We could also also perhaps detect the overflow and replace with `len1`. It seems that in `js_of_ocaml` `Sys.max_string_length + Sys.max_string_length < 0`.

That is replace  

```
Int.min len1 (i + limit - 1)
```
by 
```
let max = 
  let max = Int.min len1 (i + limit - 1)
  if max < 0 then len1 else max 
```
But it's a bit more comparisons in the loop.

dbuenzli replied:
So I went with that there's no measurable effect.

Octachron commented on stdlib/string.ml:
```diff
@@ -280,3 +280,100 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+  let ignore =
+    (* Value used to make the values around the diagonal stripe ignored
+       by the min computations when we have a limit. *)
+    limit + 1
```
Using `limit` is sufficient, since this is already the maximal element of the implicit distance type.

dbuenzli replied:
Indeed but it's easier to follow what happens when you debug with a "don't care value".

Octachron reviewed (commented):
Sorry for the staggered review, but I have one more comment about the diagonal stripe width and another about overflow for `js_of_ocaml`.

dbuenzli committed:
String.edit_distance: reduce window size and add comment about rogue read.

dbuenzli committed:
String.edit_distance: jmax computation: guard against reasonable overflows.

Octachron commented on stdlib/string.ml:
```diff
@@ -326,7 +326,11 @@ let edit_distance' ?(limit = Int.max_int) s (s0, len0) s1 =
-    for j = Int.max 1 (i - limit) to Int.min len1 (i + limit - 1) do
+    let jmax =
+      let jmax = Int.min len1 (i + limit - 1) in
+      if jmax < 0 then (* overflow *) len1 else jmax
```
I have an alternative computation using `0 < i` and `0 < limit ≤ len1`  with one less comparison and one more subtraction:
```ocaml
let jmax =
   let im = i - 1 in
   if im > len1 - limit then len1 else im + limit
```

dbuenzli replied:
I think I rather keep the obvious formulation. FWIW I did measure the change with the new overflow detection and there was no noticable peformance change.

Octachron replied:
This is fine with me.

Octachron commented on stdlib/string.ml:
```diff
@@ -280,3 +280,108 @@ let get_int32_le s i = B.get_int32_le (bos s) i
+          (* Note when j = i - limit, the latter [row] read makes a bogus read
+             on the value that was in the matrix at d.(i-2).(i - limit - 1).
+             Since by induction for all i,j, d.(i).(j) >= abs (i - j),
+             (row.(j-1) + 1) is greater or equal to [limit]. *)
```
The sentence feels incomplete, I would add "and the bogus read is thus ignored by the min" or similar to close the reasoning.

dbuenzli committed:
Tweak comment.

Octachron commented on Changes:
```diff
@@ -128,11 +128,18 @@ Working version
    approximating unicode width.
+
+- #13570, Format: add an out_width function to Format device for approximating
+   unicode width.
```
There is conflict resolution duplication (but I can fix it later).

dbuenzli replied:
Ah sorry yes. Did it on the way.

Octachron reviewed (approved):
The current implementation state looks good to me, thanks for enduring my nitpicking !

dbuenzli committed:
Revert changes to .gitattributes.

They are no longer needed since #13724 has been merged.

Octachron commented on testsuite/tests/lib-string/test_string.ml:
```diff
@@ -76,3 +76,59 @@ let () =
+  let max_dist s = if String.length s <= 1 then 1 else 2 in
+  test ~max_dist ["abc"] "a" [];
+  test ~max_dist ["abc"; "ab"; "b"] "a" ["ab"; "b"];
+  ()
```
This file lacks a `non-ascii` attribute in https://github.com/ocaml/ocaml/blob/trunk/.gitattributes .

dbuenzli committed:
Add non-ascii to .gitattributes apparently #13724 was not a revolution.

dbuenzli committed:
Fix conflict resolution bug in Changes.

dbuenzli committed:
.gitattributes: thinking too much about types.
